tutorials/haskell/haskell-time-tutorial/tutorial.md

```diff
@@ -0,0 +1,537 @@
+---
+title: Time, patterns and a little more
+published: 2021-11-25
+tags: haskell, pattern-synonyms, property-testing
+libraries: time
+language: haskell
+author: Felix Miño
+author-name: Felix Miño
+github-profile: felixminom
+description: In this tutorial we’re going to explore the use of pattern synonyms in Haskell, later we’ll give a brief overview to the Stack Builders’ contribution to the `time` Haskell library and finally we’ll review how to use the utilities that were introduced in the contribution.
+---
+
+As Stack Builders one of the core values of our company is contributing
+to open source (OSS). We believe that we have the power to change
+people’s lives (or at least make it easier) by pushing the boundaries
+of the software industry and we’re happy to announce that the past month
+we were able to contribute to one of Haskell's core libraries,
+[time][time-hackage].
+
+We started with the idea: “What if we bring some utilities that Ruby on
+Rails (RoR) has to Haskell?”. For instance, “How can I get the range of
+days that a month has?” or “How can I get a quarter’s boundaries?” were
+some of the questions that came to our mind. In the following example
+I will show the Haskell code needed to get all days in a month and its
+version in RoR.
+
+```ruby
+today = Date.today
+today.next_month.all_month
+=> Wed, 01 Nov 2021..Fri, 30 Nov 2021
+```
+
+```haskell
+allMonth :: Day -> [Day]
+allMonth day =
+    let (y, m , _) = toGregorian day
+    in [fromGregorian y m 1 .. fromGregorian y m 31]
```

Comment:
Would using `YearMonthDay` work here? I don't have the appropriate version of `time` to test this:

```haskell
allMonth (YearMonthDay y m _) = [YearMonthDay y m 1 .. YearMonthDay y m 31]
```

If it works, I don't think it should be added here. Instead, it could be shown after talking about patterns and `YearMonthDay` (it might be even better than the one with the new type class).

Reply:
Didn't get this one, should I change it or not ? It will work with `YearMonthDay` btw.

Reply:
I think it shouldn't be changed here because `YearMonthDay` hasn't been introduced yet, but it might be good to add after patterns and `YearMonthDay` are discussed if you want (ignoring the 31, it looks like a good alternative implementation to me).

Reply:
Yes I'm using this pattern in some places after it's introduced. I think the intention of this snippet is also to highlight that previous to our contribution the code needed was not pretty clear and not very readable, so I think in this case the 31 plays as a good example.